[do not merge] apply scientific-python template to enable CI/CD

ebrahimebrahim opened the pull request:
This PR is not meant to be merged but just to help explore/discuss the changes needed to apply the [scientific-python template](https://github.com/scientific-python/cookie) to HD_BET. If successful, the actual PR will be on the original repository and not this fork.


- is the move from `HD_BET` to `src/hd_bet` okay?
- we will probably have to turn off static type checking

ebrahimebrahim committed:
apply scientific-python template to enable CI/CD

ebrahimebrahim committed:
add pre-commit dependency

ebrahimebrahim committed:
adapt entrypoint to the new project organization

ebrahimebrahim committed:
make automated style fixes using ruff

ebrahimebrahim committed:
remove codespell from pre-commit

ebrahimebrahim commented:
@jcfr I have been trying to see if I can get things working using the scientific-python template. Instead of editing code manually to fix all the linting errors, I want to 

- accept the style fixes that can be done automatically by ruff (already done in this PR)
- configure the linting to ignore any remaining linting rules that are broken

The latter step I am not sure how to do. I don't see any pylint or ruff configuration file where I can tell it what rules to ignore. Can you point me in the right direction? Thanks!!

jcfr commented:
To exclude specific issues you could do the following. See https://github.com/ImagingDataCommons/idc-index/blob/abcc8550aeec97705249d430b862496c628a11b1/pyproject.toml#L152-L176

jcfr commented:
Instead of https://github.com/brain-microstructure-exploration-tools/HD-BET/pull/1/commits/477a7a36e2cbb3a194631d18aeb7df70d532b51c, you may want to apply these changes before the transition like it is done in https://github.com/ImagingDataCommons/idc-index/commit/ca633aa53dc8bce34befa891209e6bb3a4574671

jcfr commented:
It may be sensible to have one first commit only running `end-of-file-fixer`, `mixed-line-ending`, `trailing-whitespace`, `black`, `blacken-docs` and `mirrors-prettier` pre-commit hooks, and perform the additional changes in a follow-up commit.

ebrahimebrahim commented:
Thanks for the advice on this!

ebrahimebrahim committed:
ignore linting constraints that are initially not met

ebrahimebrahim committed:
style: use import rather than from-import syntax

Not doing this was resulting in a mysterious pylint error:

E0611: No name 'resize' in module 'skimage.transform' (no-name-in-module)

Not sure why this was happening

ebrahimebrahim committed:
set CD to publish to pypi instead of testpypi

ebrahimebrahim committed:
remove pypy from CI

jcfr commented on .pre-commit-config.yaml:
```diff
@@ -46,11 +46,6 @@ repos:
         args: ["--fix", "--show-fixes"]
       - id: ruff-format
 
-  - repo: https://github.com/codespell-project/codespell
```
codespell is usually a "low maintenance" dependency, was the amount of false positive too large ?

ebrahimebrahim replied:
I think I just didn't want to do a bunch of spelling corrections -- the main goal was to get the package published. But I guess if it makes future maintenance easier I can bring back codespell and do the corrections

jcfr commented on LICENSE:
```diff
@@ -198,4 +198,4 @@
    See the License for the specific language governing permissions and
-   limitations under the License.
\ No newline at end of file
+   limitations under the License.
```
In the commit message associated with `make automated style fixes using ruff`, consider listing the fixes that were applied.

For example:
* https://github.com/Slicer/Slicer/pull/7332
* https://github.com/Slicer/Slicer/pull/7647

jcfr commented on pyproject.toml:
```diff
@@ -123,6 +123,35 @@ ignore = [
   "PLR09",    # Too many <...>
   "PLR2004",  # Magic value used in comparison
   "ISC001",   # Conflicts with formatter
+  "EXE002",   # The file is executable but no shebang is present
```
```suggestion
  # Exceptions below are specific to this project
  "EXE002",   # The file is executable but no shebang is present
```

ebrahimebrahim commented:
Another attempt is now at #2
